tests/integration/test_compat_mode.py

```diff
@@ -0,0 +1,125 @@
+# THIS FILE IS PART OF THE CYLC WORKFLOW ENGINE.
+# Copyright (C) NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Tests for Cylc 7 compatibility mode."""
+
+from typing import TYPE_CHECKING
+
+from cylc.flow.cycling.integer import IntegerPoint
+from cylc.flow.data_store_mgr import TASK_PROXIES
+
+if TYPE_CHECKING:
+    from cylc.flow.scheduler import Scheduler
+
+
+async def test_blocked_tasks_in_n0(flow, scheduler, run, complete):
+    """Ensure that tasks with no satisfied dependencies remain in the pool.
+
+    In this example, the "recover" task is not satisfiable because its upstream
+    dependency "foo:failed" will never be satisfied. The unsatisfiable
+    "recover" task should remain in n=0 until removed/completed.
+
+    See https://github.com/cylc/cylc-flow/issues/4983
+    """
+    id_ = flow(
+        {
+            'scheduling': {
+                'initial cycle point': '1',
+                'cycling mode': 'integer',
+                'runahead limit': 'P2',
+                'dependencies': {
+                    'P1': {
+                        'graph': '''
+                            foo:fail => recover
+                            foo | recover => bar
+                        ''',
+                    },
+                },
+            },
+        },
+        filename='suite.rc',
+    )
+    schd: 'Scheduler' = scheduler(id_, paused_start=False, debug=True)
```

Comment:
I'm a little surprised that `debug=True` isn't the default. Is there a reason why it's not?